src/Servers/Kestrel/perf/Microbenchmarks/Http2/FlowControl/InputFlowControl.cs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.FlowControl;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Microbenchmarks.Http2.FlowControl;
+
+public class FlowControlBenchmark
+{
+    private readonly InputFlowControl _flowControl = new(1000, 10);
+    private const int N = 100000;
+    private const int Spin = 50;
+
+    [IterationSetup]
+    public void IterationSetup()
+    {
+        _flowControl.Reset();
+    }
+
+    [Benchmark]
+    public async Task ThreadsAdvanceWithWindowUpdates()
+    {
+        _flowControl.Reset();
+        var t1 = Task.Factory.StartNew(() =>
+        {
+            for (int i = 0; i < N; i++)
+            {
+                _flowControl.TryUpdateWindow(16, out _);
+            }
+            _flowControl.Abort();
+        }, TaskCreationOptions.LongRunning);
+
+        var t2 = Task.Factory.StartNew(() =>
+        {
+            for (int i = 0; i < N; i++)
+            {
+                if (_flowControl.TryAdvance(1))
+                {
+                    for (int j = 0; j < Spin; j++)
+                    {
+                    }
+                }
+            }
+        }, TaskCreationOptions.LongRunning);
+
+        var t3 = Task.Factory.StartNew(() =>
+        {
+            for (int i = 0; i < N; i++)
+            {
+                if (_flowControl.TryAdvance(1))
+                {
+                    for (int j = 0; j < Spin; j++)
+                    {
+                    }
+                }
+            }
+        }, TaskCreationOptions.LongRunning);
+
+        var t4 = Task.Factory.StartNew(() =>
+        {
+            for (int i = 0; i < N; i++)
+            {
+                if (_flowControl.TryAdvance(1))
+                {
+                    for (int j = 0; j < Spin; j++)
+                    {
+                    }
+                }
+            }
+        });
+        await Task.WhenAll(t1, t2, t3, t4);
```

Comment:
Could we make the number of parallel loops calling `TryUpdateWindow` a parameter/const like `N`?We really only need one loop calling `TryAdvance` since that represents data coming serially in over a single HTTP/2 connection. The `TryUpdateWindow` loops represent the up to 100 concurrent streams consuming the data on the thread pool.

Reply:
Updated the Benchmark as suggested and run the performance tests again (before and after changes):

BEFORE:

```
|                          Method |     Mean |    Error |   StdDev |  Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ThreadsAdvanceWithWindowUpdates | 75.01 ms | 1.471 ms | 2.972 ms | 13.33 |     - |     - |     - |      4 KB |
```

```
|                          Method |     Mean |    Error |   StdDev |  Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ThreadsAdvanceWithWindowUpdates | 73.27 ms | 1.465 ms | 3.594 ms | 13.65 |     - |     - |     - |      3 KB |
```

```
|                          Method |     Mean |    Error |   StdDev |  Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ThreadsAdvanceWithWindowUpdates | 86.33 ms | 1.682 ms | 2.667 ms | 11.58 |     - |     - |     - |      3 KB |
```

AFTER:

```
|                          Method |     Mean |    Error |   StdDev |   Median |  Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |---------:|---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ThreadsAdvanceWithWindowUpdates | 23.74 ms | 2.717 ms | 8.011 ms | 19.71 ms | 42.13 |     - |     - |     - |      4 KB |
```

```
|                          Method |     Mean |    Error |   StdDev |   Median |  Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |---------:|---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ThreadsAdvanceWithWindowUpdates | 23.73 ms | 2.831 ms | 8.346 ms | 18.65 ms | 42.13 |     - |     - |     - |      3 KB |
```

```
|                          Method |     Mean |    Error |   StdDev |  Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ThreadsAdvanceWithWindowUpdates | 26.97 ms | 3.037 ms | 8.954 ms | 37.08 |     - |     - |     - |      3 KB |
```

Reply:
Moved to https://github.com/dotnet/aspnetcore/pull/57968